### DISCUSSION TITLE
Feature request: quick copy-paste

### DISCUSSION THREAD

**Discussion by ocalhoun6** [OpenRCT2/OpenRCT2#20450] [Feature requests]
Inspired by the copying feature in Mindustry: https://youtu.be/bSr5OcI0_JU?t=906 (Ignore most of the video -- relevant part starts at 15:06 and goes for just a few seconds; the feature is used again at 15:41) Mindustry is open source as well, so you can look at how they implemented it ... though their code base is surely very different, so it won't be that relevant.

In Mindustry, you can hold down a shortcut key (they use F) and then drag your mouse over a group of buildings. As soon as you release the shortcut key, you're instantly set up to start building copies of whatever you just selected, now in a different spot. It also presents a button for that 'schematic' to be saved for later use. (The copied design can then be rotated and mirrored with other shortcut keys. They use mouse wheel for rotation and Z for mirroring.)

It occurred to me that something like this would be _great_ for OpenRCT2. A tool (with hotkey) for quickly selecting groups of rides and/or scenery, and being able to paste down more copies of them. While this could be useful for saving track/scenery designs, to be used in conjunction with an improved track/scenery design saving system and file format, you don't always want to permanently save what you're copying. Often, the part you're copying will only be relevant to the park you're currently working on, and you only need it to be 'saved' for a few moments.

-----

For example, if you've built half of a building and want the other half to be symmetrical, you could copy the first half, rotate it 180 degrees, and then paste it right down next to the first half, and then *poof*, your building is instantly completed. (But maybe it's not a building you'll ever want to build again, so you don't want to save the design permanently.)

Or for another example, you're cheesing a scenario and spamming microcoasters, but this scenario doesn't let you use your favorites, so you're stuck with something like a micro Virginia Reel or something. So you build a group of 4 of those with paths around them, then copy the whole group and paste it a few times. (But as long as you never come up against a scenario like that again, you won't want to have the design saved permanently.)

Or maybe you've built one half of a mirrored pair of dueling coasters, complete with paths and scenery, and now you want to copy it all over, mirrored, for the other half? (But it's not that great of a design -- you don't want to save it forever and use it in other scenarios, you just want to copy it just this once.)

Or maybe you want to build two paths in parallel, but not touching each other? Well, you could copy a short section of it, and then quickly paste it over and over again to build your long path that way (without needing to save something as stupidly simple as two path tiles as a saved design).

Of course, if you *did* want to save the design permanently, that should be just an easy click away after you've already copied it.

This allows for a very simple and intuitive system for saving track/scenery designs ... with an extra twist that it now becomes super-easy to copy and paste things within a scenario without needing to save them permanently and clutter up your list of saved designs with things that won't be useful in other scenarios.

------

For an OpenRCT2 implementation, it should probably be implemented as a separate tool (much like the bulldozer tool). Select that tool, and a tiny option window opens, giving you a few options like checkboxes for rides, scenery, paths, terrain -- to control just what you are and aren't copying. (With sensible defaults, of course.) The window should probably also have a button for 'clear selection' and 'save selection'. Maybe also buttons for rotate/mirror, in case people don't know the shortcut keys. 'new selection' and 'save selection' should be greyed out until the player has actually made a selection.

Also when you open the tool, the cursor changes and you can drag to select an area on the map (similar to dragging to select an area in cut-away view).

Once you've selected something, you get a ghost of your selected ride/scenery to be placed. Placing it should not remove the ghost -- that way you can quickly paste many copies of it if you want. Or it can be rotated or mirrored, or canceled by using the tool options window. Or the selection could be saved as a saved track/scenery design.

Maybe the tool window could even have a 'load saved selection' button to allow placing saved designs?

------

I know this wouldn't be easy at all to implement, and it also depends on other things like a revamped track/scenery saving system and format -- one that can save things that include multiple rides or no rides at all.

I just want to put this out there as an idea for how that new system could work -- a way to make it so quick and easy to use that players could really come to depend on it.

**Comment by J-PIE-314** [comment]
The ability to copy and paste multiple items has been implemented in this plug-in: https://github.com/Sadret/openrct2-scenery-manager.
